### PR TITLE
added extractors

### PR DESCRIPTION
**Pull request by hexated**
- add Embedgram
- add Mvidoo
- add Streamplay
- add FEnet
- fix UpstreamExtractor

### CODE REVIEW

File: app/src/main/java/com/lagradost/cloudstream3/extractors/Embedgram.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        val quality = document.select("video source:last-child").attr("title")
+        callback.invoke(
+            ExtractorLink(
+                this.name,
+                this.name,
+                httpsify(link),
+                "$mainUrl/",
+                getQualityFromName(quality),
+                headers = mapOf(
+                    "Range" to "bytes=0-"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/Embedgram.kt:32]: 🤨 does not exoplayer take care of this?

File: app/src/main/java/com/lagradost/cloudstream3/extractors/Mvidoo.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        Regex("source\\s*src=\"([^\"]+)").find(data)?.groupValues?.get(1)?.let { link ->
+            callback.invoke(
+                ExtractorLink(
+                    this.name,
+                    this.name,
+                    link,
+                    "$mainUrl/",
+                    Qualities.Unknown.value,
+                    headers = mapOf(
+                        "Range" to "bytes=0-"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/Mvidoo.kt:41]: ?

File: app/src/main/java/com/lagradost/cloudstream3/extractors/Streamplay.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                        this.name,
+                        res.file ?: return@map null,
+                        "$mainServer/",
+                        when (res.label) {
+                            "HD" -> Qualities.P720.value
+                            "SD" -> Qualities.P480.value
+                            else -> Qualities.Unknown.value
+                        },
+                        headers = mapOf(
+                            "Range" to "bytes=0-"
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/extractors/Streamplay.kt:67]: ?

**Review by LagradOst — APPROVED** [review]

File: app/src/main/java/com/lagradost/cloudstream3/extractors/Embedgram.kt (hunk 1 above)
Comment (hexated) [reply app/src/main/java/com/lagradost/cloudstream3/extractors/Embedgram.kt:32]: idk, but it doesn't work without that, same case with [Egybest](https://github.com/ImZaw/cloudstream-extensions-arabic/blob/9b2fe358a2ef7c3a4946b30c9ec5f2dd1609f98b/EgyBestProvider/src/main/kotlin/com/egybest/EgyBestProvider.kt#L255-L256)

**Pull request merged**